// angular2-mqtt/index.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { NgModule } from '@angular/core';
import { MqttService } from './src/mqtt.service';
export * from './src/mqtt.service';
export * from './src/mqtt.model';
export var MQTT_SERVICE_OPTIONS = {
    connectOnCreate: true,
    hostname: 'localhost',
    port: 1884,
    path: ''
};
export function mqttServiceFactory() {
    return new MqttService(MQTT_SERVICE_OPTIONS);
}
var MqttModule = MqttModule_1 = (function () {
    function MqttModule() {
    }
    MqttModule.forRoot = function (providedService) {
        if (providedService === void 0) { providedService = {
            provide: MqttService,
            useFactory: mqttServiceFactory
        }; }
        return {
            ngModule: MqttModule_1,
            providers: [providedService]
        };
    };
    return MqttModule;
}());
MqttModule = MqttModule_1 = __decorate([
    NgModule({})
], MqttModule);
export { MqttModule };
var MqttModule_1;
